charmcraft/application/commands/store.py

```diff
@@ -0,0 +1,1999 @@
+# Copyright 2020-2023 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For further info, check https://github.com/canonical/charmcraft
+
+"""Commands related to Charmhub."""
+import collections
+import dataclasses
+import os
+import pathlib
+import shutil
+import string
+import tempfile
+import textwrap
+import typing
+import zipfile
+from operator import attrgetter
+from typing import TYPE_CHECKING, Dict, List, Optional
+
+import yaml
+from craft_cli import ArgumentParsingError, emit
+from craft_cli.errors import CraftError
+from craft_parts import Step
+from craft_store import attenuations
+from craft_store.errors import CredentialsUnavailable
+from humanize import naturalsize
+from tabulate import tabulate
+
+import charmcraft.store.models
+from charmcraft import env, parts, utils
+from charmcraft.application.commands.base import CharmcraftCommand
+from charmcraft.models import project
+from charmcraft.store import ImageHandler, LocalDockerdInterface, OCIRegistry, Store
+from charmcraft.store.models import Entity
+from charmcraft.utils import cli
+
+if TYPE_CHECKING:
+    from argparse import ArgumentParser, Namespace
+
+
+# some types
+class _EntityType(typing.NamedTuple):
+    charm: str = "charm"
+    bundle: str = "bundle"
+
+
+class _ResourceType(typing.NamedTuple):
+    file: str = "file"
+    oci_image: str = "oci-image"
+
+
+EntityType = _EntityType()
+ResourceType = _ResourceType()
+# the list of valid attenuations to restrict login credentials
+VALID_ATTENUATIONS = {getattr(attenuations, x) for x in dir(attenuations) if x.isupper()}
+
+
+class LoginCommand(CharmcraftCommand):
+    """Login to Charmhub."""
+
+    name = "login"
+    help_msg = "Login to Charmhub"
+    overview = textwrap.dedent(
+        """
+        Login to Charmhub.
+
+        Charmcraft will provide a URL for the Charmhub login. When you have
+        successfully logged in, charmcraft will store a token for ongoing
```

Comment:
```suggestion
        successfully logged in, Charmcraft will store a token for ongoing
```

I think we use "Charmcraft" when talking about the application, and "charmcraft" mostly for command-line samples

Reply:
actually I don't know if the help text for these commands is new - let me know if you want me to review it

Reply:
It's old text, but in the vein of Continuous Improvement™, I'm always willing to pull things like that in, especially if it's a line that's already in the diff.

Well spotted!